File: work-backups/packmpi.c
/*      program packmpi.c */
/* Illustrates mpi_pack and mpi_unpack. */    
#include <stdio.h>
/* Includes the mpi C library. */
#include "mpi.h"
      main(int argc, char* argv[]) {
      float a,b;
      int c,d,location;
      int ierr;
      char numbers[100];
      int  my_rank,p,n,source,dest,tag,loc_n;
      int  i;
      MPI_Status status;
      n    = 4;
      dest = 0;
      tag  = 50;
/* Initializes mpi, gets the rank of the processor, my_rank, */
/* and number of processors, p. */
      MPI_Init(&argc, &argv);
      MPI_Comm_rank(MPI_COMM_WORLD,&my_rank);
      MPI_Comm_size(MPI_COMM_WORLD,&p);
/* Processor 0 packs and broadcasts the four number. */   
      if(my_rank==0) {
        a = 2.718;
        b = 3.141;
        c = 1;
        d = 186000;
        location = 0;
        MPI_Pack(&a,1,MPI_FLOAT,&numbers,100,&location,
                         MPI_COMM_WORLD);
        MPI_Pack(&b,1,MPI_FLOAT,&numbers,100,&location,
                         MPI_COMM_WORLD);
        MPI_Pack(&c,1,MPI_INT,&numbers,100,&location,
                         MPI_COMM_WORLD);
        MPI_Pack(&d,1,MPI_INT,&numbers,100,&location,
                         MPI_COMM_WORLD);
        MPI_Bcast(&(numbers[0]),100,MPI_PACKED,0,
                        MPI_COMM_WORLD);                
      }else{
        MPI_Bcast(&(numbers[0]),100,MPI_PACKED,0,
                        MPI_COMM_WORLD);
/* Each processor unpacks the numbers.  */                     
        location = 0;
        MPI_Unpack(&(numbers[0]),100,&location,&a,1,MPI_FLOAT,
                         MPI_COMM_WORLD);
        MPI_Unpack(&(numbers[0]),100,&location,&b,1,MPI_FLOAT,
                         MPI_COMM_WORLD);
        MPI_Unpack(&(numbers[0]),100,&location,&c,1,MPI_INT,
                         MPI_COMM_WORLD);
        MPI_Unpack(&(numbers[0]),100,&location,&d,1,MPI_INT,
                         MPI_COMM_WORLD);
        }
/* Each processor prints the numbers. */    
      printf("my_rank = %d a = %f\n",my_rank,a);
      printf("my_rank = %d b = %f\n",my_rank,b);
      printf("my_rank = %d c = %d\n",my_rank,c);
      printf("my_rank = %d d = %d\n",my_rank,d);
/* mpi is terminated. */          
      MPI_Finalize();
} /*     end program packmpi */

File: work-backups/packmpi.f
      program packmpi
! Illustrates mpi_pack and mpi_unpack.     
      implicit none
      include 'mpif.h'
      real:: a,b
      integer::c,d,location
      integer::ierr
      character, dimension(1:100)::numbers
      integer::  my_rank,p,n,source,dest,tag,loc_n
      integer:: i,status(mpi_status_size)
      data n,dest,tag/4,0,50/
      call mpi_init(ierr)
      call mpi_comm_rank(mpi_comm_world,my_rank,ierr)
      call mpi_comm_size(mpi_comm_world,p,ierr)
! Processor 0 packs and broadcasts the four number.   
      if (my_rank.eq.0) then
        a = exp(1.)
        b = 4*atan(1.)
        c = 1
        d = 186000
        location = 0
        call mpi_pack(a,1,mpi_real,numbers,100,location,&
                         mpi_comm_world, ierr)
        call mpi_pack(b,1,mpi_real,numbers,100,location,&
                         mpi_comm_world, ierr)
        call mpi_pack(c,1,mpi_integer,numbers,100,location,&
                         mpi_comm_world, ierr)
        call mpi_pack(d,1,mpi_integer,numbers,100,location,&
                         mpi_comm_world, ierr)
        call mpi_bcast(numbers,100,mpi_packed,0,&
                        mpi_comm_world,ierr)
      else
        call mpi_bcast(numbers,100,mpi_packed,0,&
                        mpi_comm_world,ierr)
! Each processor unpacks the numbers.                       
        location = 0
        call mpi_unpack(numbers,100,location,a,1,mpi_real,&
                         mpi_comm_world, ierr)
        call mpi_unpack(numbers,100,location,b,1,mpi_real,&
                         mpi_comm_world, ierr)
        call mpi_unpack(numbers,100,location,c,1,mpi_integer,&
                         mpi_comm_world, ierr)
        call mpi_unpack(numbers,100,location,d,1,mpi_integer,&
                         mpi_comm_world, ierr)
      end if    
! Each processor prints the numbers.     
      print*,'my_rank =',my_rank, 'a = ',a
      print*,'my_rank =',my_rank, 'b = ',b
      print*,'my_rank =',my_rank, 'c = ',c
      print*,'my_rank =',my_rank, 'd = ',d           
      call mpi_finalize(ierr)
      end program packmpi
